### ISSUE TITLE
button_down?(Gosu::GpLeft) and button_down?(Gosu::GpUp) always true w/ Gosu 0.8.x and Xbox 360 controller

### ISSUE THREAD

**Issue by brettchalupa** [jlnr/gosu#261]
I am going through the [Ruby Tutorial](https://github.com/jlnr/gosu/wiki/Ruby-Tutorial), and I tested out the player movement with different controllers, and `button_down?(Gosu::GpLeft)` is returning true, even when the controller left is not being triggered. I am on Mac OS X 10.10.1 using Gosu 0.8.6 with Ruby 2.1.3. I tried this with the PS4 controller plugged in via USB and with the Xbox 360 controller plugged in via USB (with [this driver](https://github.com/d235j/360Controller)).

I am not sure what is causing this (maybe even the code?). When I unplug either controller, the constant left rotation stops.

Here is a gif of the behavior:

![constant-left](https://cloud.githubusercontent.com/assets/928367/5600191/27d4315c-92a3-11e4-9720-92ad4ccf8c73.gif)

As you can see, it just continually rotates left.

Here is the `GameWindow#update` that is being used:

``` ruby
class GameWindow < Gosu::Window
  def update
    if button_down?(Gosu::KbLeft) || button_down?(Gosu::GpLeft)
      puts button_down?(Gosu::GpLeft)
      @player.turn_left
    end

    if button_down?(Gosu::KbRight) || button_down?(Gosu::GpRight)
      @player.turn_right
    end

    if button_down?(Gosu::KbUp) || button_down?(Gosu::GpButton0)
      @player.accelerate
    end

    @player.move
  end
end
```

The `@player` instance is just the `Player` class from the tutorial.

Let me know if there is anything I can do to help with this! Gosu has been great so far, and I am really enjoying using it. :smile:

**Comment by shawn42** [comment]
Do you have another gamepad that you could test with. I've had cheap USB gamepads flake out and get stuck buttons.

**Comment by brettchalupa** [comment]
I tested it with an Xbox 360 controller and PS4 controller, both have the same behavior. Those controllers do not experience constant left input when playing games or developing with other libraries. :sweat_smile: 

I sadly do not have another gamepad to test it with right now other than those tow.

**Comment by brettchalupa** [comment]
I tested out `button_down?(Gosu::GpUp)`, and it is also returning `true` every update loop when that is not being input. `Gosu::GpRight` and `Gosu::GpDown` work as expected.

**Comment by jlnr** [comment]
The same happened to me when I've tried the unofficial Xbox 360 driver on OS X. I am not 100% sure, but this *should* be a bug in either the 360 driver, or in SDL 2.

I wish there were more high-profile SDL 2 games, because now I don't even know what you could test to see whose fault it is. :)

**Comment by brettchalupa** [comment]
Yup, the 360 controller's stick and axis works fine in other games (Dust: An Elysian Tale, Fez, OlliOlli are some of the more recent ones).

The PS4 controller, as far as I know, is not using a that 360 driver and is experiencing the same behavior, which makes me think that it is not the controller or the driver, but one of the libraries.

**Comment by brettchalupa** [comment]
By accident, I think I may have made a little progress into figuring out why this is happening.

I was just messing around with the controller w/ plugged in and running the game, and pressing the Left Trigger (LT) inputs `GpDown` and pressing the Right Trigger (RT) inputs `GpRight`. This is the case on the Xbox 360 Controller and PS4 controller.

My assumption is that that the buttons being mapped by those trigger are:

- LT Not Pressed - `GpUp`
- LT Pressed - `GpDown`
- RT Not Pressed - `GpLeft`
- RT Pressed - `GpRight`

This would explain the reason why `GpUp` and `GpLeft` are always `true` for `#button_down?`.

If this is the case, is this at the SDL level or Gosu level?

**Comment by brettchalupa** [comment]
I have not written C++ in sometime, and I do not know much SDL, but just from poking around in the code, I wonder if the issue is from [these lines in `Input.cpp`](https://github.com/jlnr/gosu/blob/c29d7d13d866ecd143ab795b9404b5a7de3bc9d7/src/Input/Input.cpp#L80-L95)?

It seems like it loops through all of the axes on the controller, and assuming that the triggers are axes, there would be at least 4 (left - right on the stick, up - down on the stick, Left Trigger, and Right Trigger). Maybe it should only set `GpUp`, `GpLeft`, `GpDown`, `GpRight` if the axis is `0`, or `1`?

I am not sure if it is safe to assume the first two axes are the directions, but if they are, I would be happy to give submitting a patch for this a shot!

**Comment by jlnr** [comment]
Sadly, I don't think that *any* assumptions about USB gamepads are safe :) I know that I used to have a gamepad that was always reporting "top left" when it actually meant that a mini gamepad was centred, and I remember that I've fixed it in Gosu before SDL did. So, maybe this is a regression because Gosu uses an SDL backend now.

It also seems to be a bug in the 360 driver, which we might be able to fix with a Pull Request on github.

If it turns out that there is no good solution for this, I think Gosu should still check for the name of the gamepad, and only ignore some axes for these two specific controllers on OS X.

I haven't had time to connect my 360 gamepad and test this myself, will report back then :)

**Comment by jlnr** [comment]
Mini update: The Xbox 360 driver for OS X is not at fault, since the same thing happens on Windows 7.

**Comment by jlnr** [comment]
The bug didn't occur in Gosu 0.7.x *until* the user first used the left or right trigger button (axis 2 and 3), because Gosu 0.7 ignored all "axis" that have never been in a neutral state. So it is not a new bug, but it was harder to reproduce before.

**Comment by jlnr** [comment]
The correct solution is to use the new SDL_GameController API, which correctly maps the axes, hats and buttons and ships with a small database that includes the two gamepads you are using:

http://hg.libsdl.org/SDL/file/tip/src/joystick/SDL_gamecontrollerdb.h

As a kludgy workaround for now, I'll bring back the safeguard not to report any axis that has never been neutral.

**Comment by jlnr** [comment]
I've just committed GameController support: https://github.com/jlnr/gosu/commit/452f46e8f8d868c8188d26cff18518a06bf6153b

This fixes all issues with the Xbox 360 controller, but I don't have a PS4 controller for testing.

I'll release 0.8.7 ASAP :)

**Comment by jlnr** [comment]
Fixed in 0.8.7(.2) which was just released :) HTH.
